fix(ivy): ngcc - recognize synthesized constructors

JoostK opened the pull request:
ngcc needs to be able to identify synthesized constructors in order to let ngtsc create a base factory call, instead of assuming a no-args constructor.

This is important for Material, e.g. the following code ([source](https://github.com/angular/material2/blob/21e646a3e7d99323e3f6d1e2961f416c8695bf48/src/lib/table/table.ts#L27-L30)):

```typescript
export class MatTable<T> extends CdkTable<T> {
  /** Overrides the sticky CSS class set by the `CdkTable`. */
  protected stickyCssClass = 'mat-table-sticky';
}
```

is compiled to ES2015 ([source](https://unpkg.com/@angular/material@7.2.0/esm2015/table.js)):

```javascript
class MatTable extends CdkTable {
    constructor() {
        super(...arguments);
        /**
         * Overrides the sticky CSS class set by the `CdkTable`.
         */
        this.stickyCssClass = 'mat-table-sticky';
    }
}
```
and downleveled to ES5 ([source](https://unpkg.com/@angular/material@7.2.0/esm5/table.es5.js)):

```javascript
var MatTable = /** @class */ (function (_super) {
    __extends(MatTable, _super);
    function MatTable() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        /**
         * Overrides the sticky CSS class set by the `CdkTable`.
         */
        _this.stickyCssClass = 'mat-table-sticky';
        return _this;
    }
    MatTable.decorators = [/* ... */];
    return MatTable;
}(CdkTable));
```

Currently, ngcc reports the `MatTable` class to have a no-args constructor, resulting in a factory function that doesn't inject the dependencies of `CdkTable`.

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:

JoostK committed:
fix(ivy): ngcc - recognize synthesized constructors

A constructor function may have been "synthesized" by TypeScript during
JavaScript emit, in the case no user-defined constructor exists and e.g.
property initializers are used. Those initializers need to be emitted
into a constructor in JavaScript, so the TypeScript compiler generates a
synthetic constructor.

This commit adds identification of such constructors as ngcc needs to be
able to tell if a class did originally have a constructor in the
TypeScript source. When a class has a superclass, a synthesized
constructor must not be considered as a user-defined constructor as that
prevents a base factory call from being created by ngtsc, resulting in a
factory function that does not inject the dependencies of the superclass.
Hence, we identify a default synthesized super call in the constructor
body, according to the structure that TypeScript emits.

JoostK committed:
fixup! fix(ivy): ngcc - recognize synthesized constructors

JoostK committed:
fixup! fix(ivy): ngcc - recognize synthesized constructors

JoostK commented:
@petebacondarwin I have updated the code with some more documentation and added ES2015 support. For the ES2015 tests I choose to include them in the existing file, whereas I split out the ES5 constructor tests into a separate file. I am now wondering which is better 🤔

Regarding the tests, I feel it's a bit overkill to add tests that verify all possible mismatches in AST structure. How do you feel about that? Are the current tests sufficient or do you see some additional ones that should be added?

petebacondarwin commented on packages/compiler-cli/src/ngcc/test/host/esm2015_host_spec.ts:
```diff
@@ -980,6 +980,69 @@ describe('Fesm2015ReflectionHost', () => {
+        const file = {
+          name: '/synthesized_constructors.js',
+          contents: `
+            class TestClass {
```
Do we need to have a super class here?

JoostK replied:
It won't affect the tests, but let's include it for correctness.

JoostK replied:
Done!

petebacondarwin reviewed (approved):
LGTM

petebacondarwin commented on packages/compiler-cli/src/ngcc/src/host/esm5_host.ts:
```diff
@@ -254,6 +258,146 @@ function reflectArrayElement(element: ts.Expression) {
+}
+
+function isSuperIdentifier(expression: ts.Expression): boolean {
+  return ts.isIdentifier(expression) && expression.text.startsWith('_super');
```
why `startsWith` rather than `===`?

JoostK replied:
They may become `_super_1` to avoid duplicate names. There's a test for that.

I felt like a regex is overkill as we don't check that the identitier is used consistently across all usages anyways, so `startWith` should be good enough.

petebacondarwin replied:
Let's go with this for now - please can you add a comment to explain why you're using `startsWith`.
I am a bit worried about false-positives but I expect they would be very rare...

JoostK replied:
Done

petebacondarwin commented:
I would be tempted to keep the tests in the files rather than splitting them out. It makes the git history a bit simpler.

JoostK committed:
fixup! fix(ivy): ngcc - recognize synthesized constructors

JoostK committed:
fixup! fix(ivy): ngcc - recognize synthesized constructors

JoostK committed:
fixup! fix(ivy): ngcc - recognize synthesized constructors

JoostK committed:
fixup! fix(ivy): ngcc - recognize synthesized constructors

JoostK commented:
I added an integration test in b29dd5ceda6469f10caa6b026657f96a1e7cf3da only to find that ES5 didn't actually work :-)

I added fixup commit 2c9df973b60ab1eaa6d917b46b39337c05ce8123 in which `hasBaseClass` is specialized for ES5, that _should_ fix the issue.

The test still fails though, as we're now emitting `const` into an ES5 file. This is mostly due to the same issue as we recently had with emitting `!` into JavaScript files. There's a few options to solve this:

1. Always translate declarations as `var` instead of `const`.
2. Let the transformer accept a `mode` such that it can decide how to emit a declaration.
3. Avoid the Angular to TS translation and emit from the Angular AST.

JoostK committed:
fixup! fix(ivy): ngcc - recognize synthesized constructors

mhevery commented:
http://test/OCL:228414744:BASE:228414779:1546991340803:5880acf7

kara commented:
@JoostK Can you rebase this one?

JoostK commented:
@kara I can't at the moment, I'm on holiday this week. Will rebase coming weekend if none of you have beaten me to it (the conflict is trivial to resolve, only a Js docblock was added)

petebacondarwin commented:
I have rebased on master. This has triggered the code-owners approval process.
